src/cargo/ops/cargo_package.rs

```diff
@@ -309,62 +381,104 @@ fn build_ar_list(
         let readme_path = Path::new(readme);
         let abs_file_path = paths::normalize_path(&pkg.root().join(readme_path));
         if abs_file_path.exists() {
-            check_for_file_and_add("readme", readme_path, abs_file_path, pkg, &mut result, ws)?;
+            check_for_file_and_add("readme", readme_path, abs_file_path, pkg, &mut files, ws)?;
         }
     }
-    result.sort_unstable_by(|a, b| a.rel_path.cmp(&b.rel_path));
 
-    Ok(result)
+    // Check that every target contains a path inside the package root
+    for (tgt, src_path) in pkg.manifest().targets().iter().filter_map(|tgt| {
+        if let TargetSourcePath::Path(path) = tgt.src_path() {
+            Some((tgt, path.as_ref()))
+        } else {
+            None
+        }
+    }) {
+        // If the target has a source outside, report it as an error
+        let abs_src_path = normalize_path(src_path);
+        if is_outsider(&abs_src_path, pkg.root()) {
+            // Custom error message for build scripts (build.rs)
+            if tgt.is_custom_build() {
+                ws.config().shell().error(&format!(
+                    "the build script doesn't appear to be inside the package root folder, modify \
+                    your package manifest (Cargo.toml) to point to a build script inside the \
+                    package"
+                ))?;
+                std::process::exit(-1);
+            }
+
+            // Default message
+            ws.config().shell().error(&format!(
+                "the source file of {0:?} target \"{1}\" doesn't appear to be a path \
+                inside of the package. It is at {2}, whereas the root the package is {3}.\
+                This may cause issue during packaging, as modules resolution and resources \
+                included via macros are often relative to the path of source files. Consider updating \
+                the path of {0:?} target \"{1}\"  to point to a path inside the \
+                root of the package to resolve this error",
+                tgt.kind(),
+                tgt.name(),
+                abs_src_path.display(),
+                pkg.root().display(),
+            ))?;
+            std::process::exit(-1);
```

Comment:
You could you [`anyhow::bail!`](https://docs.rs/anyhow/1.0/anyhow/macro.bail.html) instead of this to return a `Err` with a message.

Reply:
makes sense, I wasn't sure if the bail was going to properly do the `error: ...`